### PR TITLE
OCPQE-18400 automate OCP-56083

### PR DESCRIPTION
**Pull request by dis016**
[OCPQE-18400](https://issues.redhat.com//browse/OCPQE-18400) automate OCP-56083

### CODE REVIEW

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-4.14-upgrade-from-stable-4.13-gcp-ipi-proxy-private-f14

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.14-amd64-stable-4.14-upgrade-from-stable-4.13-gcp-ipi-proxy-private-f14

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml [hunk 1]
```diff
@@ -146,7 +146,7 @@ tests:
   steps:
     cluster_profile: gcp-qe
     env:
-      ENABLE_OTA_TEST: OCP-24358
+      ENABLE_OTA_TEST: OCP-24358,OCP-56083
```
Comment (jianlinliu) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml:149]: Per our original design, `ENABLE_OTA_TEST` should only support "false", "true", and single case id, but no multiple case id. cc @jiajliu

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 2]
```diff
@@ -685,6 +685,38 @@ function upgrade() {
         clear_upgrade
         check_upgrade_status "${cluster_src_ver}"
     fi
+
+    if check_ota_case_enabled "OCP-56083"; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:689]: From the following case steps, sounds like mainly the following steps:
1. unset channel
2. trigger upgrade
3. check clusterversion message

From my understanding, 1 should be treated as some pre-ugprade actions, 3 should be some post-upgrade actions. If that, 1 should be placed in `cucushift-ota-preupgrade`, 3 should be placed in `cucushift-ota-preupgrade`, 2 already covered in this `cucushift-upgrade-toimage`.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml (hunk 1 above)
Comment (jiajliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml:149]: Yes, correct. if OCP-56083 is not for a case which needed to be ran on a dedicate cluster with other test excluded, we could add it into run_ota_multi_test while setting ENABLE_OTA_TEST: true, which means all the cases in the caseset would be ran on the cluster.

Comment (jiajliu) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml:149]: https://github.com/openshift/release/blob/master/ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh#L284

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh (hunk 2 above)
Comment (dis016) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:689]: OK, Thanks @jianlinliu, i will update the changes accordingly

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 3]
```diff
@@ -159,6 +159,23 @@ function post-OCP-47197(){
     return 0
 }
 
+function pre-OCP-56083(){
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:162]: s/pre/post

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml (hunk 1 above)
Comment (dis016) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.14__amd64-stable-4.14-upgrade-from-stable-4.13.yaml]: Thanks @jiajliu @jianlinliu, updated in latest commit

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh (hunk 3 above)
Comment (dis016) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh]: Thanks @jiajliu corrected the typo

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 4]
```diff
@@ -792,7 +793,7 @@ do
     if ! check_signed; then
         echo "You're updating to an unsigned images, you must override the verification using --force flag"
         FORCE_UPDATE="true"
-        if check_ota_case_enabled "OCP-30832" "OCP-27986" "OCP-24358" "OCP-69968"; then
+        if check_ota_case_enabled "OCP-30832" "OCP-27986" "OCP-24358" "OCP-69968" "OCP-56083"; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:796]: For `check_ota_case_enabled`, it only works with `ENABLE_OTA_TEST: <case-id>` case.

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 5]
```diff
@@ -159,6 +159,23 @@ function post-OCP-47197(){
     return 0
 }
 
+function post-OCP-56083(){
+    echo "Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local expected_msg expected_msg
+    tmp_log=$(mktemp)
+    expected_msg="Configured channel is unset"
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:167]: Shall we check more expected message just like https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L56 for more confidence.

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 6]
```diff
@@ -159,6 +159,23 @@ function post-OCP-47197(){
...
+    oc get clusterversion -ojson | jq '.items[].status.conditions[]|select(.type == \"ReleaseAccepted\")' 2>&1 | tee "${tmp_log}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:168]: Sounds like this check is equivalent to https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L53 , shall we use a different expected message for it?

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

File: ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh [hunk 7]
```diff
@@ -792,7 +793,7 @@ do
     if ! check_signed; then
         echo "You're updating to an unsigned images, you must override the verification using --force flag"
         FORCE_UPDATE="true"
-        if check_ota_case_enabled "OCP-30832" "OCP-27986" "OCP-24358" "OCP-69968"; then
+        if check_ota_case_enabled "OCP-30832" "OCP-27986" "OCP-24358" "OCP-69968" "true"; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:796]: Hmm, I think `ENABLE_OTA_TEST: "true"` conflicts with `check_ota_case_enabled`, `ENABLE_OTA_TEST: "true"` will blindly execute all the OCP-xxx cases, while not every case need to run with a signed image.

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:796]: Sounds your case is not suitable with `ENABLE_OTA_TEST: "true"`, but `ENABLE_OTA_TEST: "OCP-XXXX"`

Comment (dis016) [reply ci-operator/step-registry/cucushift/upgrade/toimage/cucushift-upgrade-toimage-commands.sh:796]: oh ok, Thanks @jianlinliu

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 8]
```diff
@@ -224,6 +224,24 @@ function post-OCP-69948(){
...
 }
 
+function post-OCP-56083(){
+    echo "Post Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local expected_msg expected_msg
+    tmp_log=$(mktemp)
+    expected_msg='\"acceptedRisks\": \"Precondition \\\"ClusterVersionRecommendedUpdate\\\" failed because of \\\"NoChannel\\\": Configured channel is unset, so the recommended status of updating from'
+    echo "${expected_msg}"
+    oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted")' 2>&1 | tee "${tmp_log}"
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:234]: Compared with https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L59, sounds like we need take `acceptedRisks` into condieration.
```
expect("Release is accepted", get_cv_condition("ReleaseAccepted", status="True"))
risks = (get_cluster_version_history(0) or {}).get('acceptedRisks')
```
For `ReleaseAccepted`, it is gotten from `.status.conditions`, while for `acceptedRisks`, it is gotten from `.status.history`.

Comment (dis016) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:234]: Good Catch, Thanks @jianlinliu :)

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 9]
```diff
@@ -224,6 +224,30 @@ function post-OCP-69948(){
...
 
+function post-OCP-56083(){
+    echo "Post Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local expected_msg expected_msg
+    tmp_log=$(mktemp)
+    expected_msg='\"acceptedRisks\": \"Precondition \\\"ClusterVersionRecommendedUpdate\\\" failed because of \\\"NoChannel\\\": Configured channel is unset, so the recommended status of updating from'
+    echo "${expected_msg}"
+    oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted")' 2>&1 | tee "${tmp_log}" 
+    if grep -q "ReleaseAccepted" "${tmp_log}"; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]: Per https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L53, sounds like the expected message is "Release is accepted".

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh (hunk 9 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]: Oh, sorry, I miss read that, my fault, ignore my comment.

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
success run: [1767225738190131200](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/49310/rehearse-49310-periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28/1767225738190131200)

**Comment by jianlinliu** [comment]
@dis016  pls squash your commits into one, and can you also help answer https://github.com/openshift/release/pull/49310/files#r1519463591

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh (hunk 9 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:234]: Compared with https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L53, we need to ensure `ReleaseAccepted` 's `status` is `True`, right?

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 10]
```diff
@@ -278,6 +278,23 @@ function pre-OCP-69948(){
...
+
+function pre-OCP-56083(){
+    echo "Pre Test Start: OCP-56083"
+    echo "Unset the upgrade channel"
+    local expected_warn_msg expected_info_msg tmp_log
+    tmp_log=$(mktemp)
+    expected_warn_msg="warning: Clearing channel"
+    expected_info_msg="info: Cluster channel is already clear"
+    oc adm upgrade channel --allow-explicit-channel  2>&1 | tee "${tmp_log}" 
+    if grep -qe "${expected_warn_msg}\|${expected_info_msg}" "${tmp_log}"; then
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:291]: could we check clusterversion.spec directly, just like the step 2?
```
# oc get clusterversion/version -ojson | jq .spec.channel
null
```

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 11]
```diff
@@ -224,6 +224,29 @@ function post-OCP-69948(){
     return 0
 }
 
+function post-OCP-56083(){
+    echo "Post Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local expected_msg expected_msg
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:230]: dup var expected_msg?

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 12]
```diff
@@ -224,6 +224,29 @@ function post-OCP-69948(){
...
+    tmp_log=$(mktemp)
+    expected_msg='\"acceptedRisks\": \"Precondition \\\"ClusterVersionRecommendedUpdate\\\" failed because of \\\"NoChannel\\\": Configured channel is unset, so the recommended status of updating from'
+    oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted")' 2>&1 | tee "${tmp_log}"
```
Comment (jiajliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:233]: suggest to check status directly here.
```
# ./oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted").status'
True
```

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

File: ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh [hunk 13]
```diff
@@ -278,6 +278,23 @@ function pre-OCP-69948(){
...
+
+
+function pre-OCP-56083(){
+    echo "Pre Test Start: OCP-56083"
+    echo "Unset the upgrade channel"
+    local  tmp_log result
+    tmp_log=$(mktemp)
+    oc adm upgrade channel --allow-explicit-channel  2>&1 | tee "${tmp_log}" 
+    result=$(oc get clusterversion/version -ojson | jq .spec.channel)
+    if [[ "${result}" == "\"\"" ]] || [[ "${result}" == null ]] || [[ -z "${result}"  ]] || [[ "${result}" == "" ]]; then
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/preupgrade/cucushift-ota-preupgrade-commands.sh:290]:
```suggestion
    result=$(oc get clusterversion/version -ojson | jq -r '.spec.channel')
    if [[ "${result}" == null ]] || [[ -z "${result}" ]]; then
```

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 14]
```diff
@@ -224,6 +224,30 @@ function post-OCP-69948(){
...
 
+function post-OCP-56083(){
+    echo "Post Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local tmp_log expected_msg result
+    tmp_log=$(mktemp)
+    expected_msg='\"acceptedRisks\": \"Precondition \\\"ClusterVersionRecommendedUpdate\\\" failed because of \\\"NoChannel\\\": Configured channel is unset, so the recommended status of updating from'
+    result=$(oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted").status')
+    if  [[ "${result}" == "True" ]]; then
+        oc get clusterversion -ojson | jq .items[].status.history  2>&1 | tee ${tmp_log} || true
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]:
```suggestion
        oc get clusterversion -ojson | jq -r '.items[].status.history'  2>&1 | tee ${tmp_log} || true
```

Comment (JianLi-RH) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]: And I think we can only out put the last history (history[0])

Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]: Agree, that is another suggestion I talked with @dis016 , keep consistent with https://gitlab.cee.redhat.com/aosqe/v4-installer-autotest/-/blob/master/cases/TestCases/OCP-56083/upgrade_nochannel_risk.py#L59 (which is already executed for several rounds.)

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 15]
```diff
@@ -224,6 +224,30 @@ function post-OCP-69948(){
     return 0
 }
 
+function post-OCP-56083(){
+    echo "Post Test Start: OCP-56083"
+    echo "Upgrade cluster when channel is unset"
+    local tmp_log expected_msg result
+    tmp_log=$(mktemp)
+    expected_msg='\"acceptedRisks\": \"Precondition \\\"ClusterVersionRecommendedUpdate\\\" failed because of \\\"NoChannel\\\": Configured channel is unset, so the recommended status of updating from'
+    result=$(oc get clusterversion -ojson | jq -r '.items[].status.conditions[]|select(.type == "ReleaseAccepted").status')
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:233]:
```suggestion
    result=$(oc get clusterversion -ojson | jq -r '.status.conditions[]|select(.type == "ReleaseAccepted").status')
```

File: ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh [hunk 16]
```diff
@@ -224,6 +224,30 @@ function post-OCP-69948(){
...
+    if  [[ "${result}" == "True" ]]; then
+        oc get clusterversion -ojson | jq -r '.items[].status.history[0]'  2>&1 | tee ${tmp_log} || true
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/ota/postupgrade/cucushift-ota-postupgrade-commands.sh:235]:
```suggestion
        oc get clusterversion -ojson | jq -r '.items[].status.history[0].acceptedRisks' 
```
I think we can directly get the value of `acceptedRisks`, we can directly compare `expected_msg` with the output, not have to use `grep` way.

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by dis016** [comment]
/pj-rehearse periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28

**Comment by jianlinliu** [comment]
cc @liangxia to review.

**Comment by jianlinliu** [comment]
I am removing the hold label to get it merged, so that pick up this change in https://github.com/openshift/release/pull/49710

**Comment by dis016** [comment]
success run [1767968554512027648](https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/49310/rehearse-49310-periodic-ci-openshift-openshift-tests-private-release-4.15-amd64-stable-4.15-upgrade-from-stable-4.14-azure-ipi-fips-f28/1767968554512027648)

**Pull request merged**